Elasticsearch: Add approximative kNN search in addition to Cosine similarity

dadoonet opened the pull request:
The current implementation uses a slow function score query which is known to be slow because you have to iterate over all the hits to compute the score.

Elasticsearch in the recent versions, now also allow using knn search.

This PR now allows to change the configuration:

```java
// Use the Knn implementation
ElasticsearchEmbeddingStore store = ElasticsearchEmbeddingStore.builder()
  .configuration(ElasticsearchConfigurationKnn.builder().build())
  .restClient(restClient)
  .build();

// Note that this configuration is optional as it's the default behavior. So this code is equivalent:
ElasticsearchEmbeddingStore store = ElasticsearchEmbeddingStore.builder()
  .restClient(restClient)
  .build();

// Use the Scripting implementation (slower)
ElasticsearchEmbeddingStore store = ElasticsearchEmbeddingStore.builder()
  .configuration(ElasticsearchConfigurationScript.builder().build())
  .restClient(restClient)
  .build();
```

In more details, this is what this PR is bringing:

* Add a `configuration` new option in the builder. This helps to choose the implementation to use. It could be `ElasticsearchConfigurationScript` (the behavior before this change) or `ElasticsearchConfigurationKnn` for the new Knn implementation
* The Knn implementation (`ElasticsearchConfigurationKnn`) is now the default implementation
* Removes the `dimension` parameter as Elasticsearch can guess automatically this value based on the first document indexed. And anyway, it's better to let the user create an index template or the index mapping he wants.
* Modify the IT. In which case if a local instance of Elasticsearch is yet running on https://localhost:9200, we will use it by default and start Testcontainers only if needed.
* Make Elasticsearch container secured (which is closer to production use case)
* We now use the `elastic.version` `pom.xml` property to launch the right Elasticsearch TestContainer.
* Replaced the `removeAll()` implementation from a `DeleteByQuery(match all query)` to a Delete index API call which is always the recommended way to remove all documents. Note that this does not recreate an index.
* `ElasticsearchEmbeddingStoreRemoveIT` now extends the common removal test class `EmbeddingStoreWithRemovalIT`. But we need to implement in the future a `EmbeddingStoreWithRemovalIT.wait_for_ready()` method to allow even more common tests.
* Add more logs in implementation and tests

dadoonet committed:
Use knn queries in addition to script score

Note that the test part needs Java9+ but I can probably fix that in another way...

dadoonet commented:
I gave a quick look at #610 and that looks ok to me. The biggest conflict I'll have to resolve will be about IT with the new cloud tests. But I will figure this out.
So let's wait for #610 to be merged and then I will rebase my code and fix.

dadoonet commented:
I updated my branch so it's up to date with the main branch. 
But note that I modified the Java version to 9. I can change that but I'd not recommend anyone nowadays starting a new project with Java8 anyway. That's probably another discussion but I'd encourage the project to move to at least Java17 (or Java11) as a target...

Also it's not complete yet as I need to revisit the `ElasticsearchEmbeddingStoreCloudIT`. The goal is to run all the Knn and "legacy" tests whatever the env is.

dadoonet committed:
Depends on EmbeddingStoreWithFilteringIT for tests

This is failing and needs to be fixed.

Also move the cloud tests to the abstract class as we should test everything wherever Elasticsearch is running.

dadoonet committed:
Add back the filters

dadoonet committed:
Fix the Knn tests and refactoring of the toEmbeddingSearchResult method

dadoonet committed:
Remove non needed imports

dadoonet committed:
Use env variables to run the same tests on cloud, local or using test containers

dadoonet commented:
I have a question about the design. The current state of the PR introduces a new class named `ElasticsearchKnnEmbeddingStore`. With the same "semantic" as the `ElasticsearchEmbeddingStore` class.

So it's a "drop-in" replacement. From:

```java
ElasticsearchEmbeddingStore.builder()
                .restClient(restClient)
                .indexName(indexName)
                .build();
```

To:

```java
ElasticsearchKnnEmbeddingStore.builder()
                .restClient(restClient)
                .indexName(indexName)
                .build();
```

I'm wondering if @langchain4j would prefer an option within the builder. Which mean one "single class" `ElasticsearchEmbeddingStore` and then just call something like:

```java
ElasticsearchEmbeddingStore.builder()
                .restClient(restClient)
                .indexName(indexName)
                .useKnn(true)
                .build();
```

And then internally we choose the right implementation.

My guts feeling goes for the former, but I prefer to ask.

dadoonet commented:
I spoke with @swallez and he proposed another alternative which is the best world of the 2:

```java
ElasticsearchEmbeddingStore.builder()
                .restClient(restClient)
                .indexName(indexName)
                .config(XXX)
                .build();
```

Where `XXX` is a configuration class. Like `ElasticsearchKnnConfiguration` or `ElasticsearchScriptConfiguration`.

dadoonet commented:
When #610 has been added, it assumed some scores in tests when doing filtering.
The problem I have yet with the Elasticsearch Knn implementation is that Elasticsearch is adding the score of the query "match_all" (1.0) and the knn score (1.0). That gives a total score of 2.0 whereas the tests are expecting a score of 1.0.

```
java.lang.AssertionError: 
Expecting actual:
  2.0
to be close to:
  1.0
by less than 1% but difference was 100.0%.
(a difference of exactly 1% being considered valid)
```

I think we should not try to rely on the score but more on the ranking of the results.

dadoonet committed:
Add more logs

dadoonet committed:
Replace `Thread.sleep(1000)` by a `refresh()`

This is making the tests running much faster!

dadoonet committed:
Switch back to Java8

dadoonet committed:
Update comments and remove import

swallez commented:
(disclosure: I'm a co-maintainer of the Elasticsearch Java client at Elastic)

> I guess I like the idea with separate configurations in the builder.

The idea behind a config object is that ultimately we want to expose the full range of vector search features of Elasticsearch. So having a single `ElastisearchEmbeddingStore` is a way to provide a single entry point (with sane defaults), and a configuration (or a set of configuration classes) allows to provide the flexibility needed for more specific user needs.

> I am not sure I understand how/why this works like this.

About scoring, there are multiple aspects:

* approximate search is not limited to cosine distance, as there are [various ways to measure vector similarity](https://www.elastic.co/guide/en/elasticsearch/reference/current/dense-vector.html#dense-vector-params). The cosine, euclidian, dot product distances are in known ranges and can be cerced to [-1, 1]. The max_inner_product distance however doesn't have a defined range: it depends whether the vectors are normalized or not (on this topic I haven't found information about whether `EmbeddingModel` should return normalized vectors). Other distance measurements are in the works with hamming distance for 1 bit quantization.

* adding a filter can also affect the score. In Elasticsearch the filter can be used either as a filter that just restricts the search space and doesn't affect the score, or a query, which then results in hybrid scoring ([see RRF](https://www.elastic.co/guide/en/elasticsearch/reference/current/rrf.html)) and then the score isn't in a well-defined range. Maybe we should consider LangChain4J's filter as just a filter (doesn't affect the score), but ultimately we would like to have `ElasticsearchEmbeddingStore` allow hybrid search, and in that case the score value range is undefined.

dadoonet commented:
So I moved the filters within the knn query (as pre filters) and I hope that they won't affect the score anymore.
But I need https://github.com/elastic/elasticsearch-java/issues/774 to be fixed first ;)

dadoonet commented:
And I think that for the integration tests, we should either rely on the rank of each document but not on the score, or may be have a way for each IT to overload the expected value if needed as the behavior might be different depending on the embedding store implementation.

dadoonet committed:
Use knn Query

dadoonet committed:
Update to Elasticsearch 8.14.1

dadoonet committed:
Merge branch 'refs/heads/main' into use-knn

# Conflicts:
#	langchain4j-elasticsearch/src/main/java/dev/langchain4j/store/embedding/elasticsearch/ElasticsearchEmbeddingStore.java

dadoonet committed:
Update after main merge

dadoonet committed:
Merge branch 'refs/heads/main' into use-knn

dadoonet committed:
Add Elasticsearch Container and Client creation helper class

dadoonet commented:
Ha! Funny. I was working on it today. My plan is to do it by next week if possible.

dadoonet committed:
Fix most of the test failures.

`should_add_embedding_with_segment_with_metadata` is still failing.

We have a score of `0.9666443` with Elasticsearch Knn Search but we need 1.0 with 1% error (0.99).

dadoonet committed:
Merge branch 'refs/heads/main' into use-knn

dadoonet committed:
Update to Elasticsearch 8.14.3

dadoonet committed:
Fix failing test

We can't use the default Elasticsearch behavior which is `int8_hnsw` as it does some approximations and the integration tests do not allow that yet.

We force using `float` with `hnsw` instead of `int`.

dadoonet committed:
Merge branch 'refs/heads/main' into use-knn

dadoonet commented:
I fixed the failing tests. I "just" need now to do some refactoring and add a configuration as we spoke in https://github.com/langchain4j/langchain4j/pull/712#issuecomment-2066643407 ;) 

Stay tuned.

dadoonet committed:
Add a TODO about the `m` and `efConstruction`

dadoonet committed:
Use Configuration objects

With this change, the goal is to use a configuration object in the builder.
This one is responsible to execute the right implementation.

This helps to keep the existing code running and using the new standard default behavior (knn) with no code change.

dadoonet committed:
Avoid non needed changes

dadoonet committed:
removeAll should drop the index

Instead of running a very slow delete by query.

dadoonet committed:
Remove imports

dadoonet committed:
Use consistent logs

dadoonet committed:
Use the same handleBulkResponseErrors method

dadoonet committed:
Use the Elasticsearch version based on the pom.xml

dadoonet committed:
Merge branch 'refs/heads/main' into use-knn

dadoonet committed:
Make ElasticsearchEmbeddingStoreRemoveIT extends EmbeddingStoreWithRemovalIT

So we can benefit from more removal common tests

Note that this will need to implement in the future a `EmbeddingStoreWithRemovalIT.wait_for_ready()` method to allow even more common tests.

ATM I had to overwrite some existing tests because I'm missing this `wait_for_ready()` method.

dadoonet committed:
Add configuration builders

This will make this more flexible for the future when we will need to introduce some specific parameters for each implementation.

dadoonet committed:
Use abstract class instead of interface

So we can keep some methods internal only and don't expose unnecessary code.

dadoonet committed:
Remove non needed `numCandidates`

dadoonet committed:
Merge branch 'refs/heads/main' into use-knn

dadoonet commented:
Hey @langchain4j. I think this PR is now ready for review. I updated the original description.

In the future, we would also like to add more flexibility for the field names used behind the scene. With something like:

```java
ElasticsearchConfigurationKnn.builder()
   .vectorField("my_vector")
   .textField("my_text")
   .metadataField("my_metadata")
   .build()
```

But I did not want to add this in this PR.

Also, I added `restClient(restClient)` to provide a restClient instead of providing a URL, login/pwd... It's much more flexible and allows using self-signed certificates instead of disabling security for tests.

I kept the settings in the `ElasticsearchEmbeddingStore` class instead of moving everything to the new Configuration classes. I think it's ok as I'd prefer having specific settings only in `ElasticsearchConfigurationXXX` classes if any.

But I could definitely move everything to `ElasticsearchConfigurationXXX`:

```java
ElasticsearchEmbeddingStore store = ElasticsearchEmbeddingStore.builder()
  .configuration(ElasticsearchConfigurationKnn.builder().build())
  .restClient(restClient)
  .build();
```

would become:

```java
ElasticsearchEmbeddingStore store = ElasticsearchEmbeddingStore.builder()
  .configuration(ElasticsearchConfigurationKnn.builder()
      .restClient(restClient)
      .build())
  .build();
```

Let me know what you think ;)

dadoonet commented:
I just realized that I need actually to write some docs in https://github.com/langchain4j/langchain4j/blob/main/docs/docs/integrations/embedding-stores/elasticsearch.md

dadoonet committed:
Merge branch 'refs/heads/main' into use-knn

# Conflicts:
#	langchain4j-elasticsearch/src/test/java/dev/langchain4j/store/embedding/elasticsearch/ElasticsearchEmbeddingStoreCloudIT.java
#	langchain4j-elasticsearch/src/test/java/dev/langchain4j/store/embedding/elasticsearch/ElasticsearchEmbeddingStoreIT.java

dadoonet committed:
Fix import of `AllMiniLmL6V2QuantizedEmbeddingModel` class

dadoonet commented on langchain4j-elasticsearch/src/main/java/dev/langchain4j/store/embedding/elasticsearch/ElasticsearchEmbeddingStore.java:
```diff
@@ -246,23 +245,20 @@ public List<String> addAll(List<Embedding> embeddings, List<TextSegment> embedde
+            throw new ElasticsearchRequestFailedException(e.response().toString(), e);
         } catch (IOException e) {
+            log.error("[Elasticsearch encounter I/O Exception]", e);
             throw new ElasticsearchRequestFailedException(e.getMessage());
```
I have to replace this suggestion by:

```java
throw new ElasticsearchRequestFailedException(e.getMessage(), e);
```

as `ElasticsearchRequestFailedException(Exception)` does not exist.

dadoonet commented on langchain4j-elasticsearch/src/main/java/dev/langchain4j/store/embedding/elasticsearch/ElasticsearchEmbeddingStore.java:
```diff
@@ -280,41 +276,22 @@ public void removeAll(Filter filter) {
+     * of removing all documents one by one.
+     */
     @Override
     public void removeAll() {
```
There's no more `ElasticsearchKnnEmbeddingStore` but a configuration class. So the behavior for `removeAll()` is the same whatever the configuration used (knn or not). 

Side note: if you follow the Elasticsearch best practices, it's always better to have an index template which creates the desired index with the right settings/mapping when the first document is indexed.

That's why "dropping" entirely an index is ok. And when a new document is going to be added to the embeddings store, Elasticsearch will automatically create the index. No needs to keep it around. 

Another solution, if you want to mimic the previous behavior but without sacrificing the performance and the disk usage, would be to:

* drop the index (currently implemented that way)
* create a new empty index, basically it would change the code by:

```java
client.indices().delete(dir -> dir.index(indexName));
client.indices().create(dir -> dir.index(indexName));
```

But that's useless IMHO.

dadoonet commented on langchain4j-elasticsearch/src/main/java/dev/langchain4j/store/embedding/elasticsearch/ElasticsearchEmbeddingStore.java:
```diff
@@ -333,32 +310,14 @@ private void addAllInternal(List<String> ids, List<Embedding> embeddings, List<T
         }
     }
 
-    private void createIndexIfNotExist(String indexName, Integer dimension) {
```
That's right. We don't recommend creating an index from the code.
See explanations at https://github.com/langchain4j/langchain4j/pull/712/files#r1701975979

dadoonet commented on langchain4j-elasticsearch/src/test/java/dev/langchain4j/store/embedding/elasticsearch/ElasticsearchEmbeddingStoreRemoveIT.java:
```diff
@@ -8,53 +8,91 @@
-class ElasticsearchEmbeddingStoreRemoveIT {
+/**
+ * TODO add some methods like "EmbeddingStoreWithRemovalIT#wait_for_ready()"
+ * so we can remove the "specialized" implementations
```
Yeah. Do you want me to add this within this PR or as a follow up PR?

I think the later to avoid too many unrelated changes to the initial goal of this PR. But LMK. I'm happy both ways.

dadoonet committed:
Fixes after the first review:

* Put `dimension` back but as `@Deprecated`
* Throw `ElasticsearchRequestFailedException` in case of any Elasticsearch exception and remove the call to `log.error()` in that case
* Remove the `deleteIndex()` which was one single line and called only once

dadoonet committed:
Merge branch 'refs/heads/main' into use-knn

dadoonet commented:
While working on the documentation, I realized that I will need to change in a follow up PR the code in - [ElasticsearchEmbeddingStoreExample](https://github.com/langchain4j/langchain4j-examples/blob/main/elasticsearch-example/src/main/java/ElasticsearchEmbeddingStoreExample.java)

dadoonet committed:
Add documentation and deprecate using username/password/apikey

Instead of using username/password/apikey, we should have the user providing a rest client. It's much more flexible as we might have to deal with:

* provided SSL certificates
* self-signed certificates
* apiKey
* username and password although not recommended)

dadoonet commented:
@langchain4j I added also some documentation to make the Elasticsearch page a bit less "empty" ;)

dadoonet committed:
mention approximate in doc

dadoonet committed:
Add ElasticsearchRequestFailedException(Throwable)

dadoonet committed:
Merge branch 'refs/heads/main' into use-knn

dadoonet commented:
@langchain4j Could you approve the workflow so we can test the code on the CI?

dadoonet committed:
Change the default log level for tests

dadoonet commented:
I guess we are all good ;)

```
[INFO] --- failsafe:3.1.2:integration-test (default) @ langchain4j-elasticsearch ---
[INFO] Using auto detected provider org.apache.maven.surefire.junitplatform.JUnitPlatformProvider
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running dev.langchain4j.store.embedding.elasticsearch.ElasticsearchKnnEmbeddingStoreIT
15:42:49,567 INFO  [d.l.s.e.e.ElasticsearchClientHelper] Starting testcontainers with Elasticsearch [8.14.3].
15:43:03,399 INFO  [d.l.s.e.e.ElasticsearchClientHelper] Found Elasticsearch cluster version [8.14.3] running at [https://localhost:62145].
[INFO] Tests run: 103, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 30.46 s -- in dev.langchain4j.store.embedding.elasticsearch.ElasticsearchKnnEmbeddingStoreIT
[INFO] Running dev.langchain4j.store.embedding.elasticsearch.ElasticsearchEmbeddingStoreIT
15:43:19,826 INFO  [d.l.s.e.e.ElasticsearchClientHelper] Starting testcontainers with Elasticsearch [8.14.3].
15:43:31,803 INFO  [d.l.s.e.e.ElasticsearchClientHelper] Found Elasticsearch cluster version [8.14.3] running at [https://localhost:62175].
[INFO] Tests run: 103, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 23.36 s -- in dev.langchain4j.store.embedding.elasticsearch.ElasticsearchEmbeddingStoreIT
[INFO] Running dev.langchain4j.store.embedding.elasticsearch.ElasticsearchEmbeddingStoreRemoveIT
15:43:43,185 INFO  [d.l.s.e.e.ElasticsearchClientHelper] Starting testcontainers with Elasticsearch [8.14.3].
15:43:53,924 INFO  [d.l.s.e.e.ElasticsearchClientHelper] Found Elasticsearch cluster version [8.14.3] running at [https://localhost:62179].
[INFO] Tests run: 12, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 21.03 s -- in dev.langchain4j.store.embedding.elasticsearch.ElasticsearchEmbeddingStoreRemoveIT
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 218, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- failsafe:3.1.2:verify (default) @ langchain4j-elasticsearch ---
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:18 min
[INFO] Finished at: 2024-08-05T15:44:04+02:00
[INFO] ------------------------------------------------------------------------
```

dliubarskyi committed:
Merge branch 'main' into use-knn

dliubarskyi committed:
Merge branch 'main' into use-knn

dadoonet commented:
Amazing! Thanks for your patience and for reviewing! :)

dadoonet commented:
@langchain4j Could you add the `Elasticsearch` label on this one? ;)
